File: pkg-py/tests/test_handoff_data.py
import pytest
import querychat._handoff_data as handoff_data
from querychat._datasource import DataFrameSource
from querychat._handoff_types import HandoffLanguage
from querychat.data import tips


class RecordingDataFrameSource(DataFrameSource):
    def __init__(self, table_name: str):
        super().__init__(tips(), table_name)
        self.get_data_calls = 0
        self.export_error: Exception | None = None

    def get_data(self):
        self.get_data_calls += 1
        if self.export_error is not None:
            raise self.export_error
        return super().get_data()


@pytest.fixture
def tips_source():
    return DataFrameSource(tips(), "tips")


class TestHandoffDataCatalog:
    @pytest.mark.parametrize(
        ("language", "expected", "forbidden"),
        [
            ("python", "duckdb.connect()", "DBI::dbConnect"),
            ("r", "DBI::dbConnect(duckdb::duckdb())", "duckdb.connect()"),
        ],
    )
    def test_bundled_csv_instructions_match_target_language(
        self,
        tips_source: DataFrameSource,
        language: HandoffLanguage,
        expected: str,
        forbidden: str,
    ):
        catalog = handoff_data.prepare_handoff_data(
            {"tips": tips_source},
            language=language,
        )

        assert expected in catalog.prompt_instructions
        assert forbidden not in catalog.prompt_instructions

    @pytest.mark.parametrize(
        ("language", "expected", "forbidden"),
        [
            ("python", 'os.environ["DATABASE_URL"]', "Sys.getenv"),
            ("r", 'Sys.getenv("DATABASE_URL")', "os.environ"),
        ],
    )
    def test_database_instructions_match_target_language(
        self,
        language: HandoffLanguage,
        expected: str,
        forbidden: str,
    ):
        class DatabaseSource:
            def get_db_type(self) -> str:
                return "PostgreSQL"

        catalog = handoff_data.prepare_handoff_data(
            {"orders": DatabaseSource()},
            language=language,
        )

        assert expected in catalog.prompt_instructions
        assert forbidden not in catalog.prompt_instructions

    @pytest.mark.parametrize(
        ("language", "expected", "forbidden"),
        [
            ("python", 'os.environ["DATABASE_URL"]', "Sys.getenv"),
            ("r", 'Sys.getenv("DATABASE_URL")', "os.environ"),
        ],
    )
    def test_external_dataframe_instructions_require_user_supplied_setup(
        self,
        language: HandoffLanguage,
        expected: str,
        forbidden: str,
    ):
        instructions = handoff_data.external_dataframe_instructions(
            "tips",
            "DuckDB",
            language,
        )

        assert (
            'The original in-memory DataFrame for table "tips" is not bundled.'
            in instructions
        )
        assert "user-supplied data file or equivalent database connection" in instructions
        assert "dedicated DATA SETUP block" in instructions
        assert 'existing table name `"tips"`' in instructions
        assert expected in instructions
        assert forbidden not in instructions
        assert "path/to/your" not in instructions
        assert (
            "Do not claim to know the original file path, connection string, "
            "or credentials." in instructions
        )
        assert (
            "This setup may need adjustment before the handoff can run." in instructions
        )

    def test_prepare_describes_every_registered_table(self):
        sources = {
            "tips": DataFrameSource(tips(), "tips"),
            "tips_copy": DataFrameSource(tips(), "tips_copy"),
        }

        catalog = handoff_data.prepare_handoff_data(sources, language="python")

        assert set(catalog.entries) == {"tips", "tips_copy"}
        assert "tips.csv" in catalog.prompt_instructions
        assert "tips_copy.csv" in catalog.prompt_instructions

    def test_prepare_does_not_export_any_dataframe(self):
        tips_source = RecordingDataFrameSource("tips")
        unused_source = RecordingDataFrameSource("unused")

        handoff_data.prepare_handoff_data(
            {"tips": tips_source, "unused": unused_source},
            language="python",
        )

        assert tips_source.get_data_calls == 0
        assert unused_source.get_data_calls == 0

    def test_materialize_exports_only_referenced_dataframe(self):
        tips_source = RecordingDataFrameSource("tips")
        unused_source = RecordingDataFrameSource("unused")
        sources = {"tips": tips_source, "unused": unused_source}
        catalog = handoff_data.prepare_handoff_data(sources, language="python")

        context = handoff_data.materialize_handoff_data(
            catalog,
            sources,
            ["tips"],
        )

        assert set(context.bundled_files) == {"tips.csv"}
        assert context.bundled_tables == ["tips"]
        assert tips_source.get_data_calls == 1
        assert unused_source.get_data_calls == 0

    def test_materialize_deduplicates_referenced_tables_before_export(
        self,
        monkeypatch,
    ):
        source = RecordingDataFrameSource("tips")
        sources = {"tips": source}
        catalog = handoff_data.prepare_handoff_data(sources, language="python")
        csv_size = len(handoff_data.export_csv(source))
        source.get_data_calls = 0
        monkeypatch.setattr(
            "querychat._handoff_data.MAX_BUNDLE_SIZE",
            csv_size,
        )

        context = handoff_data.materialize_handoff_data(
            catalog,
            sources,
            ["tips", "tips"],
        )

        assert context.bundled_tables == ["tips"]
        assert list(context.bundled_files) == ["tips.csv"]
        assert context.data_instructions.count("tips.csv") == 1
        assert source.get_data_calls == 1

    def test_materialize_preserves_first_reference_order(self):
        sources = {
            "first": RecordingDataFrameSource("first"),
            "second": RecordingDataFrameSource("second"),
        }
        catalog = handoff_data.prepare_handoff_data(sources, language="python")

        context = handoff_data.materialize_handoff_data(
            catalog,
            sources,
            ["second", "first", "second"],
        )

        assert context.bundled_tables == ["second", "first"]
        assert list(context.bundled_files) == ["second.csv", "first.csv"]

    def test_materialized_csv_and_instructions_are_stable_after_source_mutation(self):
        source = RecordingDataFrameSource("tips")
        sources = {"tips": source}
        catalog = handoff_data.prepare_handoff_data(sources, language="python")

        context = handoff_data.materialize_handoff_data(
            catalog,
            sources,
            ["tips"],
        )
        original_csv = context.bundled_files["tips.csv"]
        original_instructions = context.data_instructions
        source._df = source._df.head(1)

        assert context.bundled_files["tips.csv"] == original_csv
        assert context.data_instructions == original_instructions

    def test_materialize_rejects_unknown_tables_before_export(self):
        source = RecordingDataFrameSource("tips")
        sources = {"tips": source}
        catalog = handoff_data.prepare_handoff_data(sources, language="python")

        with pytest.raises(handoff_data.HandoffDataError, match="unknown"):
            handoff_data.materialize_handoff_data(
                catalog,
                sources,
                ["missing"],
            )

        assert source.get_data_calls == 0

    def test_materialize_rejects_export_failures(self):
        source = RecordingDataFrameSource("tips")
        source.export_error = RuntimeError("cannot export")
        sources = {"tips": source}
        catalog = handoff_data.prepare_handoff_data(sources, language="python")

        with pytest.raises(handoff_data.HandoffDataError, match="could not export"):
            handoff_data.materialize_handoff_data(
                catalog,
                sources,
                ["tips"],
            )

        assert source.get_data_calls == 1

    def test_materialize_externalizes_all_dataframes_when_one_exceeds_limit(
        self,
        monkeypatch,
    ):
        source = RecordingDataFrameSource("tips")
        sources = {"tips": source}
        catalog = handoff_data.prepare_handoff_data(sources, language="python")
        monkeypatch.setattr("querychat._handoff_data.MAX_BUNDLE_SIZE", 1)

        context = handoff_data.materialize_handoff_data(
            catalog,
            sources,
            ["tips"],
        )

        assert context.bundled_files == {}
        assert context.bundled_tables == []
        assert context.externalized_dataframe_tables == ["tips"]
        assert "DATA SETUP" in context.data_instructions
        assert "may need adjustment" in context.data_instructions

    def test_materialize_externalizes_all_dataframes_when_combined_bundle_exceeds_limit(
        self,
        monkeypatch,
    ):
        sources = {
            "tips": RecordingDataFrameSource("tips"),
            "tips_copy": RecordingDataFrameSource("tips_copy"),
        }
        catalog = handoff_data.prepare_handoff_data(sources, language="python")
        one_table = handoff_data.materialize_handoff_data(
            catalog,
            sources,
            ["tips"],
        )
        monkeypatch.setattr(
            "querychat._handoff_data.MAX_BUNDLE_SIZE",
            len(one_table.bundled_files["tips.csv"]) + 1,
        )

        context = handoff_data.materialize_handoff_data(
            catalog,
            sources,
            ["tips", "tips_copy"],
        )

        assert context.bundled_files == {}
        assert context.bundled_tables == []
        assert context.externalized_dataframe_tables == ["tips", "tips_copy"]

File: pkg-r/tests/testthat/test-handoff_data.R
describe("prepare_handoff_data()", {
  it("classifies DataFrameSource separately from other source classes", {
    skip_if_no_dataframe_engine()
    dataframe <- local_recording_data_frame_source(engine = "sqlite")
    database <- new_fake_handoff_data_source()

    catalog <- prepare_handoff_data(
      list(tips = dataframe$source, orders = database),
      language = "python"
    )

    expect_identical(catalog$entries$tips$mode, "dataframe")
    expect_identical(catalog$entries$orders$mode, "database")
    expect_identical(catalog$entries$tips$table_name, "tips")
    expect_identical(catalog$entries$orders$db_type, "PostgreSQL")
  })

  it("classifies DBISource and TblSqlSource as database sources", {
    skip_if_not_installed("RSQLite")
    sqlite <- local_sqlite_connection(table_name = "orders")
    dbi_source <- DBISource$new(sqlite$conn, "orders")
    tbl_source <- local_tbl_sql_source(table_name = "sales")

    catalog <- prepare_handoff_data(
      list(orders = dbi_source, sales = tbl_source),
      language = "python"
    )

    expect_identical(catalog$entries$orders$mode, "database")
    expect_identical(catalog$entries$sales$mode, "database")
  })

  it("returns the final catalog shape and describes every source", {
    skip_if_no_dataframe_engine()
    first <- local_recording_data_frame_source(
      table_name = "first",
      engine = "sqlite"
    )
    second <- local_recording_data_frame_source(
      table_name = "second",
      engine = "sqlite"
    )

    catalog <- prepare_handoff_data(
      list(first = first$source, second = second$source),
      language = "r"
    )

    expect_named(
      catalog,
      c("entries", "prompt_instructions", "language"),
      ignore.order = FALSE
    )
    expect_named(catalog$entries, c("first", "second"))
    expect_identical(catalog$language, "r")
    expect_match(catalog$prompt_instructions, "first.csv", fixed = TRUE)
    expect_match(catalog$prompt_instructions, "second.csv", fixed = TRUE)
  })

  it("does not read or materialize dataframe data", {
    skip_if_no_dataframe_engine()
    used <- local_recording_data_frame_source(
      table_name = "used",
      engine = "sqlite"
    )
    unused <- local_recording_data_frame_source(
      table_name = "unused",
      engine = "sqlite"
    )

    catalog <- prepare_handoff_data(
      list(used = used$source, unused = unused$source),
      language = "python"
    )

    expect_identical(used$state$get_data_calls, 0L)
    expect_identical(unused$state$get_data_calls, 0L)
    expect_false("bundled_files" %in% names(catalog))
  })

  it("rejects unsupported languages", {
    expect_snapshot(
      error = TRUE,
      prepare_handoff_data(
        list(orders = new_fake_handoff_data_source()),
        language = "javascript"
      )
    )
  })
})

describe("render_handoff_data_instructions()", {
  it("uses target-language APIs for bundled CSV data", {
    entry <- list(
      table_name = "tips",
      db_type = "DuckDB",
      mode = "dataframe"
    )

    python <- render_handoff_data_instructions(entry, TRUE, "python")
    r <- render_handoff_data_instructions(entry, TRUE, "r")

    expect_match(python, "duckdb.connect()", fixed = TRUE)
    expect_no_match(python, "DBI::dbConnect", fixed = TRUE)
    expect_match(r, "DBI::dbConnect(duckdb::duckdb())", fixed = TRUE)
    expect_no_match(r, "duckdb.connect()", fixed = TRUE)
  })

  it("uses target-language credential APIs for database data", {
    entry <- list(
      table_name = "orders",
      db_type = "PostgreSQL",
      mode = "database"
    )

    python <- render_handoff_data_instructions(entry, FALSE, "python")
    r <- render_handoff_data_instructions(entry, FALSE, "r")

    expect_match(python, 'os.environ["DATABASE_URL"]', fixed = TRUE)
    expect_no_match(python, "Sys.getenv", fixed = TRUE)
    expect_match(r, 'Sys.getenv("DATABASE_URL")', fixed = TRUE)
    expect_no_match(r, "os.environ", fixed = TRUE)
    expect_match(python, "Do not hardcode", fixed = TRUE)
    expect_match(r, "Do not hardcode", fixed = TRUE)
  })

  it("renders safe external-data guidance for unbundled dataframes", {
    entry <- list(
      table_name = "tips",
      db_type = "DuckDB",
      mode = "dataframe"
    )

    instructions <- render_handoff_data_instructions(entry, FALSE, "r")

    expect_match(instructions, "DATA SETUP", fixed = TRUE)
    expect_match(instructions, "user-supplied", fixed = TRUE)
    expect_match(instructions, 'Sys.getenv("DATABASE_URL")', fixed = TRUE)
    expect_match(instructions, "Do not claim to know", fixed = TRUE)
  })
})

describe("materialize_handoff_data()", {
  it("rejects unknown referenced tables before exporting", {
    skip_if_no_dataframe_engine()
    fixture <- local_recording_data_frame_source(
      table_name = "tips",
      engine = "sqlite"
    )
    sources <- list(tips = fixture$source)
    catalog <- prepare_handoff_data(sources, language = "python")

    expect_snapshot(
      error = TRUE,
      materialize_handoff_data(catalog, sources, "missing")
    )
    expect_identical(fixture$state$get_data_calls, 0L)
  })

  it("deduplicates referenced tables in first-reference order", {
    skip_if_no_dataframe_engine()
    first <- local_recording_data_frame_source(
      table_name = "first",
      engine = "sqlite"
    )
    second <- local_recording_data_frame_source(
      table_name = "second",
      engine = "sqlite"
    )
    sources <- list(first = first$source, second = second$source)
    catalog <- prepare_handoff_data(sources, language = "python")

    context <- materialize_handoff_data(
      catalog,
      sources,
      c("second", "first", "second")
    )

    expect_identical(context$bundled_tables, c("second", "first"))
    expect_identical(names(context$bundled_files), c("second.csv", "first.csv"))
  })

  it("exports only referenced dataframe tables and skips database sources", {
    skip_if_no_dataframe_engine()
    dataframe <- local_recording_data_frame_source(
      table_name = "tips",
      engine = "sqlite"
    )
    unused <- local_recording_data_frame_source(
      table_name = "unused",
      engine = "sqlite"
    )
    database <- new_fake_handoff_data_source(table_name = "orders")
    sources <- list(
      tips = dataframe$source,
      unused = unused$source,
      orders = database
    )
    catalog <- prepare_handoff_data(sources, language = "python")

    context <- materialize_handoff_data(catalog, sources, c("tips", "orders"))

    expect_setequal(names(context$bundled_files), "tips.csv")
    expect_identical(context$bundled_tables, "tips")
    expect_identical(dataframe$state$get_data_calls, 1L)
    expect_identical(unused$state$get_data_calls, 0L)
  })

  it("produces UTF-8 CSV bytes with a header row", {
    skip_if_no_dataframe_engine()
    fixture <- local_recording_data_frame_source(
      data = data.frame(name = "café", stringsAsFactors = FALSE),
      table_name = "tips",
      engine = "sqlite"
    )
    sources <- list(tips = fixture$source)
    catalog <- prepare_handoff_data(sources, language = "python")

    context <- materialize_handoff_data(catalog, sources, "tips")

    csv_text <- rawToChar(context$bundled_files[["tips.csv"]])
    expect_match(csv_text, '^"name"', perl = TRUE)
    expect_match(csv_text, "café", fixed = TRUE)
  })

  it("externalizes every referenced dataframe when one export exceeds the byte budget", {
    skip_if_no_dataframe_engine()
    fixture <- local_recording_data_frame_source(
      table_name = "tips",
      engine = "sqlite"
    )
    sources <- list(tips = fixture$source)
    catalog <- prepare_handoff_data(sources, language = "python")

    context <- materialize_handoff_data(catalog, sources, "tips", max_bytes = 1)

    expect_identical(context$bundled_files, list())
    expect_identical(context$bundled_tables, character())
    expect_identical(context$externalized_dataframe_tables, "tips")
    expect_match(context$data_instructions, "DATA SETUP", fixed = TRUE)
    expect_match(context$data_instructions, "may need adjustment", fixed = TRUE)
  })

  it("externalizes every referenced dataframe when the combined budget is exceeded", {
    skip_if_no_dataframe_engine()
    first <- local_recording_data_frame_source(
      table_name = "tips",
      engine = "sqlite"
    )
    second <- local_recording_data_frame_source(
      table_name = "tips_copy",
      engine = "sqlite"
    )
    sources <- list(tips = first$source, tips_copy = second$source)
    catalog <- prepare_handoff_data(sources, language = "python")
    one_table <- materialize_handoff_data(catalog, sources, "tips")

    context <- materialize_handoff_data(
      catalog,
      sources,
      c("tips", "tips_copy"),
      max_bytes = length(one_table$bundled_files[["tips.csv"]]) + 1
    )

    expect_identical(context$bundled_files, list())
    expect_identical(context$bundled_tables, character())
    expect_identical(
      context$externalized_dataframe_tables,
      c("tips", "tips_copy")
    )
  })

  it("wraps export failures in a stable handoff-specific message", {
    skip_if_no_dataframe_engine()
    fixture <- local_recording_data_frame_source(
      table_name = "tips",
      engine = "sqlite"
    )
    fixture$state$get_data_error <- simpleError("cannot export")
    sources <- list(tips = fixture$source)
    catalog <- prepare_handoff_data(sources, language = "python")

    expect_snapshot(
      error = TRUE,
      materialize_handoff_data(catalog, sources, "tips")
    )
    expect_identical(fixture$state$get_data_calls, 1L)
  })
})
